x-pack/plugins/encrypted_saved_objects/server/lib/encrypted_saved_objects_service.ts

```diff
@@ -0,0 +1,218 @@
+   * to be encrypted if any and returns that __MUTATED__ attributes dictionary back.
+   * @param type Type of the saved object to decrypt attributes for.
+   * @param attributes Dictionary of __ALL__ saved object attributes.
+   * @param id Id of the saved object to strip encrypted attributes from.
+   * @throws Will throw if decryption fails for whatever reason.
+   * @throws Will throw if any of the attributes to decrypt is not a string.
+   */
+  public async decryptAttributes<T extends Record<string, unknown>>(
+    type: string,
+    attributes: T,
+    id: string
+  ) {
+    const typeRegistration = this.typeRegistrations.get(type);
+    if (typeRegistration === undefined) {
+      return;
+    }
+
+    const encryptionAAD = this.getAAD(typeRegistration, attributes);
+    const notDefinedAttributesToDecrypt = [];
+    for (const attributeName of typeRegistration.attributesToEncrypt) {
+      const attributeValue = attributes[attributeName];
+      if (attributeValue == null) {
+        // Normally we expect all encrypted attributes to be defined and have a `string` type, but
+        // if it's not the case we should collect and log them to make troubleshooting easier.
+        notDefinedAttributesToDecrypt.push(attributeName);
+      } else if (typeof attributeValue !== 'string') {
+        this.logger.audit.decryptAttributeFailure(attributeName, type, id);
+        throw new Error(
+          `Encrypted "${attributeName}" attribute should be a string, but found ${typeDetect(
+            attributeValue
+          )}`
+        );
+      } else {
+        try {
+          attributes[attributeName] = await this.crypto.decrypt(attributeValue, encryptionAAD);
+        } catch (err) {
+          this.logger.audit.decryptAttributeFailure(attributeName, type, id);
+          throw err;
+        }
+      }
+    }
+
+    if (notDefinedAttributesToDecrypt.length > 0) {
+      this.logger.debug(
+        `The following attributes of saved object "${type}:${id}" were expected to be decrypted, but were not defined: ${notDefinedAttributesToDecrypt}`
+      );
+    }
+
+    this.logger.audit.decryptAttributesSuccess(typeRegistration.attributesToEncrypt, type, id);
+  }
+
+  /**
+   * Generates string representation of the Additional Authenticated Data based on the specified saved
+   * object type and attributes.
+   * @param typeRegistration Saved object type registration parameters.
+   * @param attributes All attributes of the saved object instance of the specified type.
+   */
+  private getAAD<T extends Record<string, unknown>>(
+    typeRegistration: EncryptedSavedObjectTypeRegistration,
+    attributes: T
+  ) {
+    // Collect all attributes (both keys and values) that should contribute to AAD.
+    const attributesAAD = Array.from(Object.entries(attributes)).filter(([attributeKey]) => {
+      return (
+        !typeRegistration.attributesToEncrypt.has(attributeKey) &&
```

Comment:
__note:__ Not sure what was the agreement around AAD, but I'd say there is no need to have a different AAD for every encrypted attribute within the same saved object or in other words there is no need to include `encrypted-attribute#1` into AAD for `encrypted-attribute#2` and vice versa. Am I missing something?

Reply:
I believe the initial plan was to use an AAD per "object", and to store all encrypted attributes in one "attribute"; so we'd grab all attributes, JSON stringify it, and then encrypt that string and put it in a single attribute. I'm not tied to the original approach though.